Added "!bribe" command.

ridhwandll opened the pull request:
Users can now bribe an amount, to get out of the jail. Can be used once per jail period. The success chance increases the more amount user's bribe. The users can't bribe whatever amount they like, there is a max and minimum bribe amount. The admin can change the max, minimum bribe amount and success chance by giving specific command.

ridhwandll committed:
Added "!bribe" command.

0xsmft commented on src/modules/CurrencyModule.cpp:
```diff
@@ -902,7 +902,219 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+
+					});
+			});
+	//End*****************************************************************************************************************************************
```
idt we need the start and end comments?

ridhwandll replied:
That's a bribe code indicator, you can easily see that the bribe code stars where and ends where 😄 
That script is huge in size, so I added that.

0xsmft replied:
yeah but it's ez to tell imo

FancyKillerPanda replied:
I agree with beastboioioi, we don't need the start and end comments

FahimFuad replied:
Will I remove it?, if yes how?

ridhwandll replied:
Okay I removed! Thanks for the suggestions.

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -902,7 +902,219 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+			int remainingJailTime = guildList[channel.guildId()][getUserIndex(channel.guildId(), authorId)].jailTimer->remainingTime();
+			bool inJail = guildList[channel.guildId()][getUserIndex(channel.guildId(), authorId)].jailTimer->isActive();
+			bool bribeUsed = guildList[channel.guildId()][getUserIndex(channel.guildId(), authorId)].isBribeUsed;
+			auto& authorCurrency = guildList[channel.guildId()][getUserIndex(channel.guildId(), authorId)];
```
Why not use `authorCurrency` for the other three variables here (e.g. `authorCurrency.isBribeUsed`)?

You could also use the `authorCurrency` variable further below in code to avoid repeatedly finding the user through `getUserIndex()`.

Also, there is a method `getUserData()` that should be used instead of this verbose call to `operator[]` and `getUserIndex()`.

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -902,7 +902,219 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+			QRegExp re("[+]?\\d*\\.?\\d+");
+			if (!re.exactMatch(args.at(1)))
+			{
+				client.createMessage(message.channelId(), "**You can't bribe invalid amounts**");
```
I know this message is a copy from other places but I don't really like it much, it sounds a tad weird. Something like `"Your bribe amount must be valid!"` could be better maybe?

ridhwandll replied:
yeah, sure sure!

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -902,7 +902,219 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+				QString leastBribeAmount = QString::number(config.bribeLeastAmount);
+				QString output = QString(
+					":police_officer: **I don't take such a small amount!** :police_officer:\n"
+					"*(You can always give me `%1` or more %2 )*\n"
```
I think this should be ``""*(You can always give me `%1 %2` or more)*\n"``.
The symbol should be along with the amount (and get rid of the space before the closing parenthesis).

ridhwandll replied:
sure

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -902,7 +902,219 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+				client.createMessage(message.channelId(), output);
+				return;
+			}
+			if (inJail == false)
```
`if (!inJail)`

Also, if you checked this before the other `if` statements above then you could get rid of the `inJail &&` part of those conditions.

ridhwandll replied:
yeah, fixed!

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -902,7 +902,219 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+
+			if (inJail && guildList[channel.guildId()][getUserIndex(channel.guildId(), authorId)].currency() - (amountToBribe) < debtMax)
+			{
+				client.createMessage(message.channelId(), ":police_officer: **I will be loved to be bribed with that amount but are you sure that you can give me that amount?** :police_officer:");
```
This message feels a tad clunky, maybe something like `"I would love to be bribed with that much money but unfortunately you can't afford it."`

ridhwandll replied:
Cool cool, thanks!

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -902,7 +902,219 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+			std::discrete_distribution<> distribution{ { 1 - successChance, successChance } };
+			int roll = distribution(prng);
+
+			if (inJail && roll && bribeUsed == false)
```
You shouldn't need to do the additional checks here if you've returned earlier if they're `false`.

ridhwandll replied:
Will I just do... roll?

FancyKillerPanda replied:
Yeah just `if (roll)` should be fine if you’ve already returned if the other conditions weren’t true.

FancyKillerPanda commented on src/modules/CurrencyModule.cpp:
```diff
@@ -902,7 +902,219 @@ CurrencyModule::CurrencyModule(UmikoBot* client) : Module("currency", true), m_c
+				QString time = utility::StringifyMilliseconds(totalTime);
+
+				QString output = QString(
+					":police_officer: **Your bribes don't affect my loyality!** :police_officer:\n"
```
Spelling mistake (loyalty).

ridhwandll replied:
My bad, thanks!

FancyKillerPanda commented on src/modules/CurrencyModule.h:
```diff
@@ -105,6 +107,10 @@ class CurrencyModule : public Module
 		int stealVictimBonusPercent { 25 };
 		int stealFailedJailTime { 3 };
+
+		int bribeMaxAmount{ 150 };
+		int bribeLeastAmount{ 20 };
+		int bribeSuccessChance{ 69 };
```
These should be saved and loaded from the currency config file.

ridhwandll replied:
Sorry I didn't get your word's.....

FancyKillerPanda replied:
You should make it so those variables are saved to the JSON file with all the other configuration variables. They’d also be loaded at startup so their values would remain even after the bot shuts down.

ridhwandll committed:
Small Fixes and optimizations of the "!bribe" command.

ridhwandll committed:
Minor Fix at "CurrencyModule.cpp"

ridhwandll commented on src/modules/CurrencyModule.cpp:
```diff
@@ -1670,6 +1673,10 @@ void CurrencyModule::OnLoad(const QJsonDocument& doc)
 			config.stealVictimBonusPercent = serverObj["stealVictimBonusPercent"].toString("25").toUInt();
 			config.stealFailedJailTime = serverObj["stealFailedJailTime"].toString("3").toUInt();
 
+			//config.bribeSuccessChance = serverObj["bribeSuccessChance"].toString("30").toUInt();
+			//config.bribeMaxAmount = serverObj["bribeMaxAmount"].toString().toInt();
+			//config.bribeLeastAmount = serverObj["bribeLeastAmount"].toString().toInt();
+
```
Do I have to add them here?

ridhwandll committed:
The admin given commands for the bribe max, least, success get saved

ridhwandll committed:
Revert "Merge remote-tracking branch 'origin/master'"

This reverts commit 85ba46cbb14f2a18f0250765fd94427283bb4896.

ridhwandll committed:
Merge remote-tracking branch 'origin/master'

ridhwandll committed:
Merge branch 'master' into origin/master

ridhwandll committed:
Added more juice to the response of the bot regarding bribe."

ridhwandll committed:
Added "!bribe" command.


Small Fixes and optimizations of the "!bribe" command.


Minor Fix at "CurrencyModule.cpp"


The admin given commands for the bribe max, least, success get saved


Added more juice to the response of the bot regarding bribe."

ridhwandll commented:
I am doing some cleanups!
